lua/luasnip/session/snippet_collection.lua

```diff
@@ -10,6 +10,22 @@ local M = {
 	invalidated_count = 0,
 }
 
+-- depth specifies how many levels under this table should get this metamethod as well
+local function auto_creating_tables(self, key, depth)
+	local t = {}
+	if depth ~= 1 then
+		setmetatable(t, {
+			-- TODO not sure if this is that nice, creating a new function on
+			-- each time (lua-users does this by a seperate mamber)
+			__index = function(s, k)
+				return auto_creating_tables(s, k, depth - 1)
+			end,
+		})
+	end
```

Comment:
Well yes we could of course drop this restriction (as we kinda remove the metatable when we pass the tables to the outside this should be no problem). We just need to be careful in the `snippet_collection.lua` when working with these tables (e.g. `by_ft.snippets["xyz"][number]` as well as `by_prio.snippets[number]` might return unexpected empty tables (instead of `nil` or a valid snippet)).
Just noticed in fact, the depth restriction isn't in place for `by_ft` (using `0` as initial depth).

But I don't understand what you want to assert?

Reply:
> We just need to be careful in the snippet_collection.lua when working with these tables

+1, yes exactly
My idea was to `assert` this, eg. check that `depth >= 1` whenever a new `auto_creating_table` is created.
(Obviously the result is very similar to the current code, and there's no performance-benefit since we can't produce a "release"-version of the repo without asserts (though it would be interesting to look into accomplishing just that...))

Reply:
`depth >= 1` changed to `depth ~= 1` to be able to disable checking/asserting by setting depth to `0`.

Reply:
Sounds good, that might be useful in other places :+1: